### ISSUE TITLE
sql: multiple interfering jobs are created when dropping a database with intra-DB FK relationships

### ISSUE THREAD

**Issue by chrisseto** [cockroachdb/cockroach#50344]
**Describe the problem**

Dropping databases may leave orphaned table descriptors.

**To Reproduce**

```bash
docker run --rm -it cockroachdb/cockroach:v20.1.1 start-single-node --insecure
```

In a new terminal
```bash
docker exec -it $CONTAINERID
./cockroach workload init movr
./cockroach sql --insecure
```

```sql
SHOW DATABASES;
DROP database movr;
DROP database movr CASCADE;
ALTER RANGE default CONFIGURE ZONE USING gc.ttlseconds = 30;
SHOW JOBS;
```

`SHOW JOBS` will log an error
```
        job_id       |     job_type     |                                                        description                                                         | statement | user_name |  status   | running_status |             created              |             started              |             finished             |             modified             | fraction_completed |                              error                               | coordinator_id
---------------------+------------------+----------------------------------------------------------------------------------------------------------------------------+-----------+-----------+-----------+----------------+----------------------------------+----------------------------------+----------------------------------+----------------------------------+--------------------+------------------------------------------------------------------+-----------------
  564709885874470913 | SCHEMA CHANGE GC | GC for DROP DATABASE movr CASCADE                                                                                          |           | root      | failed    | NULL           | 2020-06-17 15:03:00.60133+00:00  | 2020-06-17 15:03:00.609372+00:00 | 2020-06-17 15:03:32.161086+00:00 | 2020-06-17 15:03:30.607766+00:00 |                  0 | attempting to GC tables: fetching table 53: descriptor not found |              1
```

**Expected behavior**
I expected the GC job to succeed.

**Additional data / screenshots**

Logs and debug.zips from real world cases:
* https://github.com/cockroachlabs/support/issues/473
* https://github.com/cockroachlabs/support/issues/486

**Additional context**
The largest issue I've found is that this can prevent full cluster restores from working.

A job that consistently fails _may_ consume excess memory.

**Comment by yuzefovich** [comment]
cc @spaskob since you're L2

**Comment by ajwerner** [comment]
Let's also loop @pbardea in here for his domain knowledge if he doesn't mind. As far as I can tell, we create GC jobs for the tables as well as the database, no? If so, then maybe this is okay. If no, then maybe this is muy mal. 

It seems like that the zone config update is causing problems for our GC job.

**Comment by ajwerner** [comment]
Also @chrisseto thanks for the easy repro of the issue!

**Comment by ajwerner** [comment]
Okay so the bug seems to be that somehow we're both creating GC jobs for each table and creating a GC job for the `DROP DATABASE`.

**Comment by ajwerner** [comment]
I'm relatively sure that this has to do with foreign keys. We create a separate drop job as we cascade the table drops, hence the weird `GC for` descriptions. 

I think there are two fixes here, one for backport, which is to allow the GC job for the database to continue despite failing to find a table and the other is to not create these jobs when dropping the database

**Comment by ajwerner** [comment]
@lucy-zhang prepare yourself for some sadness.

When we drop this database, we end up create a job for the referenced table descriptor. The idea with this `InvalidMutationID` just to wait for leases. This happens here:
https://github.com/cockroachdb/cockroach/blob/50ab5362171501f4714a835817616d77b2f3061f/pkg/sql/drop_table.go#L499

This is also sort of fine and good except maybe a little bit weird because we're dropping this referenced table too. The place where this all explodes is here:

https://github.com/cockroachdb/cockroach/blob/50ab5362171501f4714a835817616d77b2f3061f/pkg/sql/schema_changer.go#L436

In this portion of the schema changer we assume that it was this job that is responsible for the dropping of this table and thus we create a GC job. We, however, have already created a schema change job for the `DROP DATABASE` that's going to create a GC job for all of the tables in the database. This turns out to be problematic because they can't both succeed. This is especially problematic for the drop database job because if it fails, it won't GC the data for the other tables. GC jobs failing is a pretty gigantic deal because it leads to orphaned data :(.

-----

To fix:

1) Stop freaking out in the GC job, just mark tables you can't find as having been GC'd. This can get backported. The UX around all of this is pretty garbage anyway.

2) Stop creating these extra jobs for tables we're dropping in this statement. This is harder. I wonder if there are any problems with multiple drops in different statements in the same transaction.

**Comment by lucy-zhang** [comment]
Thanks for looking into this. To start off, I'm going to send out a PR to do (1) soon, unless @pbardea has any objections to this approach.

**Comment by pbardea** [comment]
Thanks for looking into it - no objections here. Looking back, there shouldn't be many reasons that should cause the GC job to fail.

**Comment by lucy-zhang** [comment]
Looking at the code, I'm also concerned about the fact that the entire job will also fail if either clearing table data or deleting the descriptor fails for a single table. But it's less clear what to do in those cases. Presumably we'd want to have the job ultimately fail, but still try to GC the other tables.

Maybe what we should do is to have `OnFailOrCancel` continue to attempt to the GC the remaining tables.

**Comment by pbardea** [comment]
I'm concerned what would happen if it encounters another issue during `OnFailOrCancel`. I think we should either log the errors and optionally somehow accumulate them and only really fail the job when it's done doing what it can.

**Comment by lucy-zhang** [comment]
After #50411, there are still a few other cases in the GC job where we're not handling `errDescriptorNotFound` correctly. Here, for instance: https://github.com/cockroachdb/cockroach/blob/v20.1.1/pkg/sql/gcjob/refresh_statuses.go#L126-L129
